go/arrow/csv/writer.go

```diff
@@ -0,0 +1,147 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package csv reads CSV files and presents the extracted data as records.
+package csv
+
+import (
+	"encoding/csv"
+	"fmt"
+	"github.com/apache/arrow/go/arrow"
+	"github.com/apache/arrow/go/arrow/array"
+	"io"
+)
+
+// WriterOption configures a CSV writer.
+type WriterOption func(*Writer)
```

Comment:
perhaps, instead of defining a new "WriterOption" we should keep `Option` but introduce a hidden `config` type on which the new `Option` type would operate:

```go
type Option func(*config)
type config struct {
    // ...
}
```

this would involve having to check at runtime whether the option makes sense (some write-oriented options don't make sense for a reader and vice versa...)
not sold on this `config` struct either...
what do you think?

Reply:
I feel the same way. Option need be consistent across Reader and Writer. How about we do this?
```
type Option func(interface{})

func WithComment(c rune) Option {
	return func(i interface{}) {
		r := i.(*Reader)
		r.r.Comment = c
	}
}
```

Reply:
or, perhaps:

```go
type config interface{}
type Option func(config)

type WithComma(c rune) Option {
    return func(cfg config) {
        switch cfg := cfg.(type) {
        case *Reader:
            cfg.r.Comma = c
        case *Writer:
            cfg.w.Comma = c
        default:
            panic(errors.Errorf("arrow/csv: unknown config type %T", cfg))
        }
    }
}
```

Reply:
that's better. thanks for the suggestion!